File: clojure/razor.clj
(def ^:const +expr+ '(1 2))
(def ^:const +nextpr+ '((0 1) (2 3)))

(defn interpret [e]
  (if (integer? e)
    e
    (+ (interpret (first e)) (interpret (second e)))))

(interpret +expr+)
(interpret +nextpr+)

(defn comp- [e]
  (if (integer? e)
    (list e)
    (concat (comp- (first e)) (comp- (second e)) (list :add))))

(comp- +expr+)
(comp- +nextpr+)

(defn run [ops]
  (first
    (reduce
      (fn [stack op]
        (if (integer? op)
          (cons op stack)
          (cons (+ (first stack) (second stack)) (nthrest stack 2))))
      nil ops)))

(println (run (comp- +expr+)))
(println (run (comp- +nextpr+)))
(println (-> +expr+ comp- run))
(println (-> +nextpr+ comp- run))

File: ocaml/lib/razor.ml
type expr = Val of int | Add of expr * expr

let rec walk f g = function
  | Val i -> f i
  | Add (l, r) -> g (walk f g l) (walk f g r)

let string_of_expr =
  let f i = Format.sprintf "Val %d" i
  and g l r = Format.sprintf "Add (%s, %s)" l r in
  walk f g

let interpret = walk Fun.id Int.add

type code = Push of int | AddOp

let string_of_code = function
  | Push i -> Format.sprintf "Push %d" i
  | AddOp -> "AddOp"

let compile =
  let f i = [ Push i ] and g l r = l @ r @ [ AddOp ] in
  walk f g

let run ops =
  List.hd
    (List.fold_left
       (fun stack op ->
         match (stack, op) with
         | _, Push i -> i :: stack
         | x :: y :: zs, AddOp -> (x + y) :: zs
         | _, _ ->
             let ss = String.concat ", " (List.map string_of_int stack)
             and os = String.concat ", " (List.map string_of_code ops) in
             raise
               (Failure
                  (Format.sprintf "Unexpected case; stack = %s, ops = %s" ss os)))
       [] ops)
